include/dlaf/factorization/qr/t_factor_impl.h

```diff
@@ -41,66 +49,129 @@
+    const std::size_t batch_size = workers_params.batch_size;
+    DLAF_ASSERT(workspaces.size() >= nworkers - 1, workspaces.size(), nworkers - 1);
+
+    const auto hp_scheduler = di::getBackendScheduler<Backend::MC>(thread_priority::high);
+    return ex::when_all(ex::when_all_vector(std::move(hh_tiles)), std::move(taus), std::move(tile_t),
+                        ex::when_all_vector(std::move(workspaces))) |
+           di::continues_on(hp_scheduler) |
+           ex::let_value([hp_scheduler, nworkers, batch_size](auto&& hh_tiles, auto&& taus,
+                                                              auto&& tile_t, auto&& workspaces) {
+             const std::chrono::duration<double> barrier_busy_wait = std::chrono::microseconds(1000);
+
+             return ex::just(std::make_unique<pika::barrier<>>(nworkers)) |
+                    di::continues_on(hp_scheduler) |
+                    ex::bulk(
+                        nworkers,
+                        [=, &hh_tiles, &taus, &tile_t, &workspaces](const std::size_t worker_id,
+                                                                    auto& barrier_ptr) mutable {
+                          const SizeType k = taus.get().size().rows();
+
+                          const std::size_t begin = worker_id * batch_size;
+                          const std::size_t end =
+                              std::min(worker_id * batch_size + batch_size, hh_tiles.size());
+
+                          const matrix::Tile<T, Device::CPU>& ws_worker =
+                              worker_id == 0 ? tile_t : workspaces[worker_id - 1];
+
+                          DLAF_ASSERT(equal_size(ws_worker, tile_t), ws_worker.size(), tile_t.size());
+
+                          tile::internal::set0<T>(ws_worker);
+                          lapack::lacpy(blas::Uplo::General, 1, k, taus.get().ptr(), 1, ws_worker.ptr(),
+                                        ws_worker.ld() + 1);
+
+                          // make it work on worker_id section of tiles
+                          for (std::size_t index = begin; index < end; ++index) {
+                            const matrix::Tile<const T, Device::CPU>& tile_v = hh_tiles[index].get();
+                            loop_gemv(tile_v, taus.get(), ws_worker);
+                          }
+
+                          barrier_ptr->arrive_and_wait(barrier_busy_wait);
+
+                          // reduce ws_T in tile_t
+                          if (worker_id == 0) {
+                            for (std::size_t other_worker = 1; other_worker < nworkers; ++other_worker) {
+                              tile::internal::add(T(1), workspaces[other_worker - 1], tile_t);
+                            }
+                          }
+                        }) |
+                    // Note: drop the barrier sent by the bulk and return tile_t
+                    ex::then([&tile_t](auto&&) mutable { return std::move(tile_t); });
```

Comment:
If you want to more closely match what the comment above says:
```suggestion
                    ex::then([&tile_t](auto) mutable { return std::move(tile_t); });
```
The barrier isn't releasing anything else so it's not critical to drop it as soon as possible. I just want to make it clear that with `auto&&` the barrier will not be destroyed in the lambda, but potentially much later.

I also realize now that the name `drop_value` that I've chosen for the algorithm in pika to do this is not clear enough (so much that it even misled me: https://github.com/eth-cscs/DLA-Future/pull/1257). It should likely be `ignore_value`, at least with the current semantics. If it actually consumed the values it could be called `drop_value`...

Reply:
Actually, the comment sense was more towards "ignoring" the barrier value, not in the sense of releasing it but to actually do not care about it anymore, because I just wanted to return the `tile_t` alone.

Indeed, I didn't thought about what was happening to the barrier 😅 so in this case as you said it is not going to hurt in any way probably, but I agree it is better to change as you suggested.

Thanks for pointing that out!

Reply:
see https://github.com/eth-cscs/DLA-Future/pull/1214/commits/66150a4634f1bac74d34819e2586500ae1aebed4